WIP: Add scaffolding to stand up Trillian on k8s.

vaikas opened the pull request:
Signed-off-by: Ville Aikas <vaikas@chainguard.dev>

Fixes #2672 

Here's the pieces for scaffolding and for createtree. I split them into two different directories because then applying just the scaffolding bits is easier and more well contained. Apply all the everythings :)

Createtree is a standalone job that makes creating trees in the k8s cluster easier and then because the entry is stuffed into configmap, it's easy to 'depend' on that for other jobs that may need it. For example, we have also a setup for spinning up a CTLog on top of this in sigstore, but wasn't sure if that might be of interest. I noticed that in README.md one of the steps after spinning up Trillian is to install CTLog on top of it.

https://github.com/sigstore/scaffolding/tree/main/config/ctlog

After creating the README-SCAFFOLDING.md in the examples dir, I noticed the PR template sez to add it to docs/ but I'm not sure if this belongs there since it's related to running tests, etc.

Lastly, the added kind test in github workflows only checks the configmap that the treeID was created, might be good to add some examples to test the grpc api as well, but figured that could be a followup if that makes sense.

<!---
Describe your changes in detail here.
If this fixes an issue, please write "Fixes #123", substituting the issue number.
-->

### Checklist

<!---
Go over all the following points, and put an `x` in all the boxes that apply.
Feel free to not tick any boxes that don't apply to this PR (e.g. refactoring may not need a CHANGELOG update).
If you're unsure about any of these, don't hesitate to ask. We're here to help!
-->

- [ ] I have updated the [CHANGELOG](CHANGELOG.md).
- [ ] I have updated [documentation](docs/) accordingly (including the [feature implementation matrix](docs/Feature_Implementation_Matrix.md)).

roger2hk commented:
/gcbrun

vaikas commented:
@roger2hk Thanks for the gcbrun.

trillian-pr-tests are failing but I can't see why, can somebody share what the problem is so I can try to fix it :) 
https://github.com/google/trillian/pull/2754/checks?check_run_id=6774522500

pavelkalinnikov commented on go.mod:
```diff
@@ -16,7 +16,7 @@ require (
 	github.com/go-redis/redis v6.15.9+incompatible
 	github.com/go-sql-driver/mysql v1.6.0
-	github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
+	github.com/golang/glog v1.0.0
```
There is a bunch of updates in `go.mod` that seem unrelated. Could you minimise the number of new/modified dependencies, or confirm that every change here is necessary?

vaikas replied:
Yeah, I just ran go mod tidy, so guess it updated some other deps. I'll take a peeksie to see what's up. My gut feeling is that mayhaps it's the k8s deps. That glog looks pretty old though if I'm reading things correctly?

pavelkalinnikov replied:
I think we can upgrade `glog` in a separate PR, it looks reasonable.

pavelkalinnikov replied:
@vaikas What's your Go tooling version? It looks like it deletes some `// indirect` deps that our tooling did not, so maybe we just have different versions.

vaikas replied:
Yeah, i am running with 1.18, so that may be it. IIRC there's a flag to go mod compatibilty. I'll look and retry.

vaikas replied:
Done, I'll resolve this and if something is still amiss, please reopen and I'll fix.

vaikas replied:
Ok. I'll cut a different PR from the head that _only_ does the go mod changes, correct?

pavelkalinnikov replied:
Thank you!

mhutchinson reviewed (commented):
I've left a few nits from a first glance. Do you want to rebase this onto HEAD, and meanwhile this week I'll try setting up an environment so I can follow the instructions here to really make sure it makes sense to me :-)

(apologies for late response, life has been busy the last 3 weeks!)

vaikas commented:
Thanks for the comments! Sorry for tardy response, was out on PTO, will pick this back up this week! @mhutchinson totes understood about life sometimes getting very busy :)

AlCutter commented:
/gcbrun

vaikas commented on examples/deployment/kubernetes/README-SCAFFOLDING.md:
```diff
@@ -0,0 +1,62 @@
+  * log-signer
+  * log-server
+  * mysql server
+2. Let's make sure everything comes up ready:
```
Fantastic, thank you so much for cutting your fingers :)  I'll tear down my local stuff, and redo from scratch!

mhutchinson commented:
/gcbrun

mhutchinson commented:
/gcbrun

mhutchinson reviewed (commented):
LGTM, modulo a few comments.

vaikas commented on examples/deployment/kubernetes/scaffolding/log-server/placeholder.go:
```diff
@@ -0,0 +1,15 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trillianlogserver
```
Great catch. This is a leftover from a while ago, when I had to import things for reasons that I can't recall anymore. Removed.

vaikas commented on .github/workflows/kind-scaffolding-e2e-createtree.yml:
```diff
@@ -0,0 +1,76 @@
+    - name: Install yq
+      uses: mikefarah/yq@70403375d7b96075bd68b40c434807cff1593568 # v4.25.1
+
+    - name: Setup mirror
+      uses: chainguard-dev/actions/setup-mirror@main
+      with:
+        mirror: mirror.gcr.io
```
In some cases esp. in test environments this can reduce flakiness by not hitting rate limits on pulling containers. So by default we set this up before having to add it later :)

mhutchinson commented:
/gcbrun

mhutchinson commented:
This is failing running the lint checks:
```
Already have image (with digest): gcr.io/trillian-opensource-ci/trillian_testbase
running gofmt
running goimports
running golangci-lint
level=warning msg="[runner] Can't run linter goanalysis_metalinter: buildir: failed to load package version: could not load export data: no export data for \"sigs.k8s.io/release-utils/version\""
level=error msg="Running error: 1 error occurred:\n\t* can't run linter goanalysis_metalinter: buildir: failed to load package version: could not load export data: no export data for \"sigs.k8s.io/release-utils/version\"\n\n"
```

mhutchinson commented:
https://github.com/golangci/golangci-lint/discussions/1920 may be relevant. Holler if you can't work this out and I'll take a deeper look.

mhutchinson commented:
/gcbrun

mhutchinson commented:
This is still causing some problem with the linter. Fortunately it can be reproduced locally:

```
golangci-lint version
golangci-lint has version v1.47.3 built from (unknown, mod sum: "h1:ri7A2DgtFpxgqcMSsU3qIT0IBm/SCdYgXlvmJx4szUU=") on (unknown)

golangci-lint run
WARN [runner] Can't run linter goanalysis_metalinter: buildir: failed to load package v1: could not load export data: no export data for "k8s.io/client-go/applyconfigurations/meta/v1"
ERRO Running error: 1 error occurred:
	* can't run linter goanalysis_metalinter: buildir: failed to load package v1: could not load export data: no export data for "k8s.io/client-go/applyconfigurations/meta/v1"
```

mhutchinson commented:
I think this is because client-go is min version go1.18: https://github.com/kubernetes/client-go/blob/master/go.mod

https://github.com/golangci/golangci-lint/discussions/1920#discussioncomment-2578691

mhutchinson commented:
Hah, I've just spent ages messing about with looking into the linters and go versions. The answer appears to be as simple as running go build to find the root cause:

```
go build ./...
# k8s.io/client-go/applyconfigurations/meta/v1
../../go/pkg/mod/k8s.io/client-go@v0.23.8/applyconfigurations/meta/v1/unstructured.go:64:38: cannot use doc (variable of type *"github.com/googleapis/gnostic/openapiv2".Document) as type *"github.com/google/gnostic/openapiv2".Document in argument to proto.NewOpenAPIData
```

mhutchinson commented:
Root cause is probably https://github.com/kubernetes/client-go/issues/1084. @vaikas can I leave this with you?

vaikas commented:
Thank you so much @mhutchinson for digging into this. I'll keep looking at it, oh what joy! 🤣

vaikas commented:
Ok, added a replace to go.mod and now things are building:
```
vaikas@villes-mbp trillian % go mod tidy -go=1.17 -compat=1.17
vaikas@villes-mbp trillian % git status
On branch k8s-createtree
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   go.mod
	modified:   go.sum
no changes added to commit (use "git add" and/or "git commit -a")
vaikas@villes-mbp trillian % go build ./...
vaikas@villes-mbp trillian %
```
Let's try this now :)

mhutchinson commented:
/gcbrun

mhutchinson commented:
Error is

```
Already have image (with digest): gcr.io/trillian-opensource-ci/trillian_testbase
running gofmt
running goimports
running golangci-lint
level=warning msg="[runner] Can't run linter goanalysis_metalinter: buildir: failed to load package version: could not load export data: no export data for \"sigs.k8s.io/release-utils/version\""
level=error msg="Running error: 1 error occurred:\n\t* can't run linter goanalysis_metalinter: buildir: failed to load package version: could not load export data: no export data for \"sigs.k8s.io/release-utils/version\"\n\n"
```

Looks like a similar issue to the last one but with a different module. Oddly though, I can't reproduce this one locally with go build or golangci-lint. Will have a quick poke at this, but due to RL considerations I may need to leave this for @AlCutter to take over next week.

JAORMX commented:
@vaikas if you want to revisit this PR, the cloudbuild results are now viewable so it would be easier to debug.

vaikas commented:
oooooh, yes, thank you for the pointer, I'll try to carve some time in the
next couple of days :)

On Mon, Dec 12, 2022 at 10:30 PM Juan Antonio Osorio <
***@***.***> wrote:

> @vaikas <https://github.com/vaikas> if you want to revisit this PR, the
> cloudbuild results are now viewable so it would be easier to debug.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/google/trillian/pull/2754#issuecomment-1347810939>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACWB45GGBHAOLK7P5B3TB6DWNAJX3ANCNFSM5YCUENJA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mhutchinson commented:
I'm back! Getting this PR in would be a great accomplishment for my return if you're up for it @vaikas :-)

vaikas commented:
Welcome back, happy new year!!!! Yes, I'll make time for it, this time I
really mean it.

On Mon, Jan 9, 2023 at 2:04 PM Martin Hutchinson ***@***.***>
wrote:

> I'm back! Getting this PR in would be a great accomplishment for my return
> if you're up for it @vaikas <https://github.com/vaikas> :-)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/google/trillian/pull/2754#issuecomment-1375524623>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACWB45CMHJYWF46YALXJZCTWRP5EBANCNFSM5YCUENJA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

vaikas commented:
@mhutchinson would you mind hitting the magic test button (/gcbrun) for me please? :)

roger2hk commented on .github/workflows/kind-scaffolding-e2e-createtree.yml:
```diff
@@ -0,0 +1,76 @@
+    - uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b # v2.4.0
+    - uses: actions/setup-go@b22fbbc2921299758641fab08929b4ac52b32923 # v2.2.0
+      with:
+        go-version: '1.17'
```
Please update the Go version to 1.19.

vaikas replied:
lol, ofc. Thanks, guess the world has moved on since I originally opened this PR :)

roger2hk commented on .github/workflows/kind-scaffolding-e2e-createtree.yml:
```diff
@@ -0,0 +1,76 @@
+        go-version: '1.17'
+        check-latest: true
+
+    - uses: imjasonh/setup-ko@2c3450ca27f6e6f2b02e72a40f2163c281a1f675 # v0.4
```
Consider upgrade to ko-build/setup-ko@v0.6.

vaikas replied:
Ditto :)

roger2hk commented:
/gcbrun

JAORMX commented on .github/workflows/kind-scaffolding-e2e-createtree.yml:
```diff
@@ -0,0 +1,76 @@
+      matrix:
+        k8s-version:
+        - v1.22.x
+        - v1.23.x
```
Any reason not to add newer kube versions?

vaikas replied:
LOL, oh dear :) Yes, done.

roger2hk commented:
/gcbrun
